pkg/envoy/ads/stream.go

```diff
@@ -246,22 +248,24 @@ func respondToRequest(proxy *envoy.Proxy, discoveryRequest *xds_discovery.Discov
 		return false
 	}
 
-	// Nonces match
-	// At this point, there is no error and nonces match, it is guaranteed an ACK with last sent version.
+	// At this point, there is no error and nonces match. It can either be an ACK or envoy could still be
+	// requesting a different set of resources on the current version version for non-wildcard TypeURIs.
 	proxy.SetLastAppliedVersion(typeURL, requestVersion)
 
-	// ----
-	// What's left is to check if the resources listed are the same. If they are not, we must respond
-	// with the new resources requested.
-	//
+	// Update subscribed resources for this TypeURI with the ones on this request.
 	// In case of LDS and CDS, "Envoy will always use wildcard mode for Listener and Cluster resources".
-	// The following logic is not needed (though correct) for LDS and CDS as request resources are also empty in ACK case.
-	//
+	// The following logic is not needed (though correct) for both LDS and CDS as both requested and last sent resources
+	// will purposefully remain as empty sets at all times, and could be avoided altogether.
+	resourcesRequested := getRequestedResourceNamesSet(discoveryRequest)
+	proxy.SetSubscribedResources(typeURL, resourcesRequested)
```

Comment:
If this can be avoided for LDS and CDS, what's the motivation to execute it?

Reply:
I had removed it initially, but decided to leave it as is to keep smaller amount of changes possible and check with reviewers first. Glad we think alike here.

I had something around the lines of:
```
if IsWildCardTypeURI(typeuri) {
   // skip resource names checks, treat as ACK (which it is)
}
```

Reply:
If skipping for wildcard type is correct, might be a good idea to do it. If this doesn't work, then our understanding of the protocol can be corrected.

Up to you on this one.